File: apps/api/services/organization_bootstrap.py
from __future__ import annotations

import re
import secrets
import string
from typing import Annotated, Literal

from fastapi import Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from core.config import Settings, get_settings
from core.errors import AppHTTPException
from core.id import generate_id, normalize_slug
from core.logging import get_logger
from core.org_permissions import OWNER_ROLE_NAME
from core.timestamps import now_unix_seconds
from db.models import Organization
from db.repositories.memberships import MembershipRepository
from db.repositories.organizations import OrganizationRepository
from db.repositories.users import UserRepository
from db.session import get_db
from providers.protocol import AuthProvider
from providers.workos.adapter import get_auth_provider
from services.provisioning import provision_organization

_SLUG_PATTERN = re.compile(r"^[a-z0-9][a-z0-9-]*[a-z0-9]$")
_SLUG_MAX_LENGTH = 64
_SLUG_COLLISION_LIMIT = 50
_RANDOM_SUFFIX_LENGTH = 6
_RANDOM_SUFFIX_ALPHABET = string.ascii_lowercase + string.digits

logger = get_logger(__name__)


class OrganizationBootstrapService:
    """Creates organizations and their owner membership for existing users."""

    def __init__(self, provider: AuthProvider, db: AsyncSession) -> None:
        self._provider = provider
        self._db = db
        self._memberships = MembershipRepository(db)
        self._organizations = OrganizationRepository(db)
        self._users = UserRepository(db)

    async def resolve_registration_slug(self, name: str, slug: str | None) -> str:
        return await self._resolve_slug(name, slug, error_domain="auth")

    async def bootstrap_existing_user(
        self,
        *,
        owner_user_id: str,
        name: str,
        slug: str | None,
    ) -> Organization:
        user = await self._users.get_by_id(owner_user_id)
        if not user:
            raise AppHTTPException(
                code="user/not-found",
                message="No user exists with the provided identifier.",
                http_status_code=status.HTTP_404_NOT_FOUND,
            )

        organization_name = name.strip()
        if not organization_name:
            raise AppHTTPException(
                code="organization/validation-failed",
                message="Organization name is required.",
                http_status_code=status.HTTP_400_BAD_REQUEST,
            )
        resolved_slug = await self._resolve_slug(organization_name, slug, error_domain="organization")
        organization_id = generate_id("organization")
        workos_organization_id: str | None = None

        try:
            workos_org = await self._provider.create_organization(
                name=organization_name,
                external_id=organization_id,
                metadata={"slug": resolved_slug, "owner_workos_user_id": user.workos_user_id},
            )
            workos_organization_id = workos_org["id"]
            workos_membership = await self._provider.create_organization_membership(
                user_id=user.workos_user_id,
                organization_id=workos_organization_id,
                role_slug="admin",
            )

            now = now_unix_seconds()
            organization = await self._organizations.create(
                id=organization_id,
                workos_organization_id=workos_organization_id,
                name=organization_name,
                slug=resolved_slug,
                status="active",
                metadata_=workos_org.get("metadata"),
                created_at=now,
                updated_at=now,
            )
            org_roles = await provision_organization(self._db, organization.id, now)
            owner_role = org_roles.get(OWNER_ROLE_NAME)

            await self._memberships.create(
                id=generate_id("membership"),
                organization_id=organization.id,
                user_id=user.id,
                workos_membership_id=workos_membership["id"],
                role=OWNER_ROLE_NAME,
                role_id=owner_role.id if owner_role else None,
                status="active",
                created_at=now,
                updated_at=now,
            )
            return organization
        except Exception:
            if workos_organization_id is not None:
                try:
                    await self._provider.delete_organization(organization_id=workos_organization_id)
                    logger.info(
                        "organization.bootstrap_existing_user.compensated",
                        workos_organization_id=workos_organization_id,
                    )
                except Exception:
                    logger.warning(
                        "organization.bootstrap_existing_user.compensation_failed",
                        workos_organization_id=workos_organization_id,
                        exc_info=True,
                    )
            raise

    async def _resolve_slug(
        self,
        name: str,
        slug: str | None,
        *,
        error_domain: Literal["auth", "organization"],
    ) -> str:
        explicit_slug = slug.strip() if slug else ""
        if not explicit_slug:
            return await self._generate_unique_org_slug(name)

        if not (3 <= len(explicit_slug) <= _SLUG_MAX_LENGTH and _SLUG_PATTERN.fullmatch(explicit_slug)):
            raise AppHTTPException(
                code="auth/invalid-input" if error_domain == "auth" else "organization/validation-failed",
                message="Please check your input." if error_domain == "auth" else "Invalid organization slug.",
                http_status_code=status.HTTP_400_BAD_REQUEST,
            )

        if await self._organizations.get_by_slug(explicit_slug):
            raise AppHTTPException(
                code=(
                    "auth/organization-slug-taken"
                    if error_domain == "auth"
                    else "organization/duplicate-slug"
                ),
                message=(
                    "This organization slug is already taken."
                    if error_domain == "auth"
                    else "An organization with this slug already exists."
                ),
                http_status_code=status.HTTP_409_CONFLICT,
            )
        return explicit_slug

    async def _generate_unique_org_slug(self, name: str) -> str:
        base = normalize_slug(name)[:_SLUG_MAX_LENGTH].rstrip("-")
        if len(base) < 3:
            base = "workspace"

        if not await self._organizations.get_by_slug(base):
            return base

        for index in range(2, _SLUG_COLLISION_LIMIT + 1):
            candidate = self._append_slug_suffix(base, str(index))
            if not await self._organizations.get_by_slug(candidate):
                return candidate

        for _ in range(_SLUG_COLLISION_LIMIT):
            suffix = "".join(secrets.choice(_RANDOM_SUFFIX_ALPHABET) for _ in range(_RANDOM_SUFFIX_LENGTH))
            candidate = self._append_slug_suffix(base, suffix)
            if not await self._organizations.get_by_slug(candidate):
                return candidate

        raise AppHTTPException(
            code="organization/slug-generation-failed",
            message="Could not generate a unique organization slug.",
            http_status_code=status.HTTP_409_CONFLICT,
        )

    @staticmethod
    def _append_slug_suffix(base: str, suffix: str) -> str:
        prefix = base[: _SLUG_MAX_LENGTH - len(suffix) - 1].rstrip("-")
        return f"{prefix}-{suffix}"


def get_organization_bootstrap_service(
    db: Annotated[AsyncSession, Depends(get_db)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> OrganizationBootstrapService:
    return OrganizationBootstrapService(provider=get_auth_provider(settings), db=db)


OrganizationBootstrapServiceDep = Annotated[
    OrganizationBootstrapService,
    Depends(get_organization_bootstrap_service),
]

File: apps/api/src/services/organization-bootstrap.ts
import type {
  ProvisioningSelectionContext,
  ProvisioningSetupSelection,
} from '@876/core/types/provisioning-selection'

import { getSettings } from '@/config'
import { AppHttpError } from '@/http/errors'
import { generateId, normalizeSlug } from '@/platform/ids'
import { getLogger } from '@/platform/logger'
import { SUPER_ADMIN_ROLE_NAME } from '@/platform/permissions'
import { nowUnixSeconds } from '@/platform/timestamps'
import { getAuthProvider } from '@/providers/workos/adapter'

import * as repository from './organization-bootstrap.repository'
import type {
  MembershipRow,
  OrganizationRow,
  UserRow,
} from './organization-bootstrap.repository'
import {
  resolveInitialProvisioningSelection,
  type ProvisioningWorkspaceDefaults,
} from './provisioning-policy'
import { workspace } from './workspace'

const log = getLogger('organization-bootstrap')

const SLUG_PATTERN = /^[a-z0-9][a-z0-9-]*[a-z0-9]$/
const SLUG_MAX_LENGTH = 64
const SLUG_COLLISION_LIMIT = 50
const RANDOM_SUFFIX_LENGTH = 6
const RANDOM_SUFFIX_ALPHABET = 'abcdefghijklmnopqrstuvwxyz0123456789'

export type OrganizationBootstrapProvider = {
  createOrganization(params: {
    name: string
    externalId: string
    metadata: Record<string, string>
  }): Promise<{ id: string; metadata?: unknown }>
  createOrganizationMembership(params: {
    userId: string
    organizationId: string
    roleSlug: string
  }): Promise<{ id: string }>
  deleteOrganization(organizationId: string): Promise<void>
}

export type OrganizationBootstrapRepository = {
  findUserById(userId: string): Promise<UserRow | null>
  findOrganizationBySlug(slug: string): Promise<OrganizationRow | null>
  createOrganization(data: {
    id: string
    workosOrganizationId: string | null
    name: string
    slug: string
    status: string
    countryCode: string | null
    currencyCode: string
    language: string
    provisioningSetupKey: string
    provisioningSelectionType: 'policy' | 'fallback'
    provisioningMatchGroupKey: string | null
    provisioningMatchPriority: number | null
    provisioningMatchedFields: string[]
    provisioningSetupSelectedAt: bigint
    metadata: unknown
    createdAt: bigint
    updatedAt: bigint
  }): Promise<OrganizationRow>
  createMembership(data: {
    id: string
    organizationId: string
    userId: string
    workosMembershipId: string | null
    role: string
    roleId: string | null
    status: string
    createdAt: bigint
    updatedAt: bigint
  }): Promise<MembershipRow>
}

export type OrganizationBootstrapWorkspace = {
  setup(
    organizationId: string,
    options?: {
      sourceAppId?: string | null
      finance?: 'ready' | 'defer'
      requireProvisioningSelection?: boolean
      now?: number
    }
  ): Promise<Record<string, { id: string }>>
  finance: {
    ensure(params: { organizationId: string }): Promise<unknown>
  }
  work: {
    ensure(params: { organizationId: string }): Promise<void>
  }
}

export type OrganizationBootstrapDeps = {
  provider: OrganizationBootstrapProvider
  repository: OrganizationBootstrapRepository
  workspace: OrganizationBootstrapWorkspace
  resolveProvisioning(context: Partial<ProvisioningSelectionContext>): Promise<{
    selection: ProvisioningSetupSelection
    defaults: ProvisioningWorkspaceDefaults
  }>
}

export function appendSlugSuffix(base: string, suffix: string): string {
  const prefix = base
    .slice(0, SLUG_MAX_LENGTH - suffix.length - 1)
    .replace(/-+$/g, '')
  return `${prefix}-${suffix}`
}

function randomSuffix(): string {
  let out = ''
  for (let i = 0; i < RANDOM_SUFFIX_LENGTH; i++) {
    const idx = Math.floor(Math.random() * RANDOM_SUFFIX_ALPHABET.length)
    out += RANDOM_SUFFIX_ALPHABET[idx]!
  }
  return out
}

export async function generateUniqueOrgSlug(
  repository: OrganizationBootstrapRepository,
  name: string
): Promise<string> {
  let base = normalizeSlug(name).slice(0, SLUG_MAX_LENGTH).replace(/-+$/g, '')
  if (base.length < 3) base = 'workspace'

  if (!(await repository.findOrganizationBySlug(base))) return base

  for (let index = 2; index <= SLUG_COLLISION_LIMIT; index++) {
    const candidate = appendSlugSuffix(base, String(index))
    if (!(await repository.findOrganizationBySlug(candidate))) return candidate
  }

  for (let i = 0; i < SLUG_COLLISION_LIMIT; i++) {
    const candidate = appendSlugSuffix(base, randomSuffix())
    if (!(await repository.findOrganizationBySlug(candidate))) return candidate
  }

  throw new AppHttpError({
    code: 'organization/slug-generation-failed',
    message: 'Could not generate a unique organization slug.',
    httpStatus: 409,
  })
}

async function resolveSlug(
  repository: OrganizationBootstrapRepository,
  name: string,
  slug: string | null | undefined,
  errorDomain: 'auth' | 'organization'
): Promise<string> {
  const explicitSlug = slug?.trim() ?? ''

  if (!explicitSlug) return generateUniqueOrgSlug(repository, name)

  if (
    explicitSlug.length < 3 ||
    explicitSlug.length > SLUG_MAX_LENGTH ||
    !SLUG_PATTERN.test(explicitSlug)
  ) {
    throw new AppHttpError({
      code:
        errorDomain === 'auth'
          ? 'auth/invalid-input'
          : 'organization/validation-failed',
      message:
        errorDomain === 'auth'
          ? 'Please check your input.'
          : 'Invalid organization slug.',
      httpStatus: 400,
    })
  }

  if (await repository.findOrganizationBySlug(explicitSlug)) {
    throw new AppHttpError({
      code:
        errorDomain === 'auth'
          ? 'auth/organization-slug-taken'
          : 'organization/duplicate-slug',
      message:
        errorDomain === 'auth'
          ? 'This organization slug is already taken.'
          : 'An organization with this slug already exists.',
      httpStatus: 409,
    })
  }

  return explicitSlug
}

export async function resolveRegistrationSlug(
  deps: OrganizationBootstrapDeps,
  name: string,
  slug: string | null | undefined
): Promise<string> {
  return resolveSlug(deps.repository, name, slug, 'auth')
}

function isRetryableWorkspaceFailure(error: unknown): boolean {
  return (
    error instanceof AppHttpError &&
    (error.code === 'provisioning/finance-workspace-unavailable' ||
      error.code === 'provisioning/work-workspace-unavailable')
  )
}

export async function bootstrapExistingUser(
  deps: OrganizationBootstrapDeps,
  params: {
    creatorUserId: string
    name: string
    slug?: string | null
    countryCode?: string | null
    subdivision?: string | null
    jurisdiction?: string | null
    /** Compatibility override; setup finance default is used when omitted. */
    currencyCode?: string | null
    /** Compatibility override; setup workspace language is used when omitted. */
    language?: string | null
    sourceAppId?: string | null
  }
): Promise<OrganizationRow> {
  const user = await deps.repository.findUserById(params.creatorUserId)
  if (!user) {
    throw new AppHttpError({
      code: 'user/not-found',
      message: 'No user exists with the provided identifier.',
      httpStatus: 404,
    })
  }

  const organizationName = params.name.trim()
  if (!organizationName) {
    throw new AppHttpError({
      code: 'organization/validation-failed',
      message: 'Organization name is required.',
      httpStatus: 400,
    })
  }

  const resolvedSlug = await resolveSlug(
    deps.repository,
    organizationName,
    params.slug ?? null,
    'organization'
  )
  const resolved = await deps.resolveProvisioning({
    country: params.countryCode ?? null,
    subdivision: params.subdivision ?? null,
    jurisdiction: params.jurisdiction ?? null,
  })

  const organizationId = generateId('organization')
  let workosOrganizationId: string | null = null

  try {
    const workosOrg = await deps.provider.createOrganization({
      name: organizationName,
      externalId: organizationId,
      metadata: { slug: resolvedSlug, owner_workos_user_id: user.workosUserId },
    })
    workosOrganizationId = workosOrg.id

    const workosMembership = await deps.provider.createOrganizationMembership({
      userId: user.workosUserId,
      organizationId: workosOrganizationId,
      roleSlug: 'admin',
    })

    const now = nowUnixSeconds()
    const nowBigint = BigInt(now)
    const selectionType = resolved.selection.match_type
    if (selectionType !== 'policy' && selectionType !== 'fallback')
      throw new Error('Initial organization selection must be policy or fallback.')

    const organization = await deps.repository.createOrganization({
      id: organizationId,
      workosOrganizationId,
      name: organizationName,
      slug: resolvedSlug,
      status: 'active',
      countryCode:
        resolved.selection.context.country ?? resolved.defaults.country_code,
      currencyCode:
        params.currencyCode?.trim().toUpperCase() ||
        resolved.defaults.currency_code,
      language: params.language?.trim() || resolved.defaults.language,
      provisioningSetupKey: resolved.selection.setup_key,
      provisioningSelectionType: selectionType,
      provisioningMatchGroupKey: resolved.selection.match_group_key,
      provisioningMatchPriority: resolved.selection.match_priority,
      provisioningMatchedFields: resolved.selection.matched_fields,
      provisioningSetupSelectedAt: nowBigint,
      metadata: (workosOrg as { metadata?: unknown }).metadata ?? null,
      createdAt: nowBigint,
      updatedAt: nowBigint,
    })

    const orgRoles = await deps.workspace.setup(organization.id, {
      sourceAppId: params.sourceAppId ?? null,
      finance: 'defer',
      requireProvisioningSelection: true,
      now,
    })
    const superAdminRole = (orgRoles as Record<string, { id: string } | undefined>)[
      SUPER_ADMIN_ROLE_NAME
    ]

    await deps.repository.createMembership({
      id: generateId('membership'),
      organizationId: organization.id,
      userId: user.id,
      workosMembershipId: workosMembership.id,
      role: SUPER_ADMIN_ROLE_NAME,
      roleId: superAdminRole?.id ?? null,
      status: 'active',
      createdAt: nowBigint,
      updatedAt: nowBigint,
    })

    await deps.workspace.finance.ensure({ organizationId: organization.id })
    await deps.workspace.work.ensure({ organizationId: organization.id })

    return organization
  } catch (error) {
    // Durable org/membership/workspace failures are retried against the persisted
    // setup. Do not delete the provider org after the local identity exists.
    if (isRetryableWorkspaceFailure(error)) throw error

    if (workosOrganizationId !== null) {
      try {
        await deps.provider.deleteOrganization(workosOrganizationId)
        log.info(
          { workos_organization_id: workosOrganizationId },
          'organization.bootstrap_existing_user.compensated'
        )
      } catch (compError) {
        log.warn(
          { err: compError, workos_organization_id: workosOrganizationId },
          'organization.bootstrap_existing_user.compensation_failed'
        )
      }
    }
    throw error
  }
}

export class OrganizationBootstrapService {
  constructor(private readonly deps: OrganizationBootstrapDeps) {}

  resolveRegistrationSlug(
    name: string,
    slug: string | null | undefined
  ): Promise<string> {
    return resolveRegistrationSlug(this.deps, name, slug)
  }

  bootstrapExistingUser(params: {
    creatorUserId: string
    name: string
    slug?: string | null
    countryCode?: string | null
    subdivision?: string | null
    jurisdiction?: string | null
    sourceAppId?: string | null
  }): Promise<OrganizationRow> {
    return bootstrapExistingUser(this.deps, params)
  }
}

function requireProviderId(
  record: Record<string, unknown>,
  resource: string
): { id: string; metadata?: unknown } {
  const id = record['id']
  if (typeof id !== 'string' || !id) {
    throw new AppHttpError({
      code: 'auth/internal-error',
      message:
        'An unexpected error occurred during authentication. Please try again later.',
      httpStatus: 502,
      description: `WorkOS returned no id for ${resource}.`,
    })
  }
  return { id, metadata: record['metadata'] }
}

export function createOrganizationBootstrapDeps(): OrganizationBootstrapDeps {
  const workos = getAuthProvider(getSettings())

  return {
    provider: {
      createOrganization: async (params) =>
        requireProviderId(
          await workos.createOrganization(params),
          'organization'
        ),
      createOrganizationMembership: async (params) =>
        requireProviderId(
          await workos.createOrganizationMembership(params),
          'organization membership'
        ),
      deleteOrganization: (organizationId) =>
        workos.deleteOrganization(organizationId),
    },
    repository,
    workspace,
    resolveProvisioning: async (context) => {
      const { selection, defaults } =
        await resolveInitialProvisioningSelection(context)
      return { selection, defaults }
    },
  }
}
